File: corerender/XTrack.cpp
//
// Created by Oogh on 2020/11/4.
//

#include "XTrack.hpp"
#include "XPlatform.hpp"
#include "XTimeline.hpp"
#include "XAudioCodec.hpp"
#include "XFFmpegVideoCodec.hpp"
#include "XIOSVideoCodec.hpp"

int XTrack::ID_GENERATOR = 0;

XTrack::XTrack() : mId(ID_GENERATOR++), mDelay(0), mClipStartTime(0), mClipEndTime(0) {
    mAudioCodec = std::make_unique<XAudioCodec>();
//    mVideoCodec = std::make_unique<XFFmpegVideoCodec>();
    mVideoCodec = std::make_unique<XIOSVideoCodec>();
}

XTrack::~XTrack() {

}

void XTrack::setTimeline(const std::shared_ptr<XTimeline>& timeline) {
    mTimeline = timeline;
}

void XTrack::setFilename(const std::string& filename) {
    mFilename = filename;

    if (mVideoCodec) {
        mVideoCodec->close();
    }
    mVideoCodec->setFilename(filename);
    mVideoCodec->open();

    if (mAudioCodec) {
        mAudioCodec->close();
    }
    mAudioCodec->setFilename(filename);
    mAudioCodec->open();
}

int XTrack::getId() const {
    return mId;
}

void XTrack::setDelay(long delay) {
    mDelay = delay;
}

void XTrack::setClipStartTime(long startTime) {
    mClipStartTime = startTime;
}

void XTrack::setClipEndTime(long endTime) {
    mClipEndTime = endTime;
}

std::string XTrack::getFilename() const {
    return mFilename;
}

long XTrack::getDelay() const {
    return mDelay;
}

long XTrack::getClipStartTime() const {
    return mClipStartTime;
}

long XTrack::getClipEndTime() const {
    return mClipEndTime;
}

long XTrack::getClipDuration() const {
    return mClipEndTime - mClipStartTime;
}

std::shared_ptr<XImage> XTrack::getImage(long clock) {
    if (mVideoCodec && (mDelay <= clock && clock <= mDelay + getClipDuration())) {
        auto result =  mVideoCodec->getImage(clock - mDelay);
        if (result) {
            result->textureId = mId;
        }
        return result;
    }
    return nullptr;
}

std::shared_ptr<XSample> XTrack::getSample(long clock, int length) {
    if (mAudioCodec && (mDelay <= clock && clock <= mDelay + getClipDuration())) {
        return mAudioCodec->getSample(length);
    }
    return nullptr;
}

File: examples/Android/XRender/corerender/src/main/java/com/core/render/XTrack.java
package com.core.render;

public class XTrack {
    private long mNativePtr = -1L;

    private XTimeline mTimeline;
    private int mId;
    private String mFilename;
    private long mDelay;
    private long mClipStartTime;
    private long mClipEndTime;

    public XTrack() {
        mNativePtr = nativeCreateTrack();
        if (mNativePtr != -1L) {
            mId = nativeGetTrackId(mNativePtr);
        }
    }

    public long getNativePtr() {
        return mNativePtr;
    }

    protected void setTimeline(XTimeline timeline) {
        mTimeline = timeline;
    }

    public void setFilename(String filename) {
        mFilename = filename;
    }

    public void setDelay(long delay) {
        mDelay = delay;
    }

    public void setClipStartTime(long clipStartTime) {
        mClipStartTime = clipStartTime;
    }

    public void setClipEndTime(long clipEndTime) {
        mClipEndTime = clipEndTime;
    }

    public int getId() {
        return mId;
    }

    public XTimeline getTimeline() {
        return mTimeline;
    }

    public String getFilename() {
        return mFilename;
    }

    public long getDelay() {
        return mDelay;
    }

    public long getClipStartTime() {
        return mClipStartTime;
    }

    public long getClipEndTime() {
        return mClipEndTime;
    }

    public void destroy() {
        if (mNativePtr != -1L) {
            nativeDestroyTrack(mNativePtr);
            mNativePtr = -1L;
        }
    }

    private native long nativeCreateTrack();
    private native void nativeSetFilename(long nativePtr, String filename);
    private native void nativeSetDelay(long nativePtr, long delay);
    private native void nativeSetClipStartTime(long nativePtr, long clipStartTime);
    private native void nativeSetClipEndTime(long nativePtr, long clipEndTime);
    private native int nativeGetTrackId(long nativePtr);
    private native void nativeDestroyTrack(long nativePtr);

}
